src/rust/cpu/cpu.rs

```diff
@@ -1928,26 +1928,20 @@ pub unsafe fn do_page_walk(
         }
 
         let kernel_write_override = !user && 0 == cr0 & CR0_WP;
-        if page_dir_entry & PAGE_TABLE_RW_MASK == 0 && !kernel_write_override && for_writing {
-            if side_effects {
-                trigger_pagefault(addr, true, for_writing, user, jit);
-            }
-            return Err(());
-        }
+        let mut allow_write = page_dir_entry & PAGE_TABLE_RW_MASK != 0;
+
+        allow_user &= page_dir_entry & PAGE_TABLE_USER_MASK != 0;
+
+        if 0 != page_dir_entry & PAGE_TABLE_PSE_MASK && 0 != cr4 & CR4_PSE {
+            // size bit is set
 
-        if page_dir_entry & PAGE_TABLE_USER_MASK == 0 {
-            allow_user = false;
-            if user {
-                // Page Fault: page table accessed by non-supervisor
+            if for_writing && !allow_write && !kernel_write_override {
```

Comment:
The user check is missing here now. I'd suggest merging it into the write check, since the body is the same (and also merge it below).

Reply:
I somehow completely forgot the user check there. These changes are now done.